### PR TITLE
Feature metrics lazyflush

### PR DESCRIPTION
**Pull request by champly**
### Issues associated with this PR

#1478 

### Solutions
You should show your solutions about the issues in your PR, including the overall solutions, details and the changes. At this time, Chinese is allowed to describe these.

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases. And you need to show the UT result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

**Comment by champly** [comment]
@nejisama @cch123 @taoyuanyuan PTAL

File: pkg/metrics/mosn.go [hunk 1]
```diff
@@ -32,8 +32,13 @@ const (
...
 )
 
-// FlushMosnMetrics marks output mosn information metrics or not, default is false
-var FlushMosnMetrics bool
+var (
+	// FlushMosnMetrics marks output mosn information metrics or not, default is false
+	FlushMosnMetrics bool
+
+	// LazyFlushMetrics marks flush data with lazy mode
+	LazyFlushMetrics bool
```
[the remark this answers is not in the document]
Comment (champly) [reply pkg/metrics/mosn.go:40]: 这个值目前没法设置，和上面的FlushMosnMetrics一样，也可以弄成配置文件里面

File: pkg/metrics/store.go [hunk 2]
```diff
@@ -153,7 +153,14 @@ func (s *metrics) Counter(key string) gometrics.Counter {
 		return gometrics.NilCounter{}
 	}
 
-	return s.registry.GetOrRegister(key, shm.NewShmCounterFunc(s.fullName(key))).(gometrics.Counter)
+	construct := func() gometrics.Counter {
```
Comment (taoyuanyuan) [inline pkg/metrics/store.go:156]: if LazyFlushMetrics {
   return
}
return s.registry.GetOrRegister(key, shm.NewShmCounterFunc(s.fullName(key))).(gometrics.Counter)

File: pkg/metrics/store.go [hunk 3]
```diff
@@ -162,7 +169,14 @@ func (s *metrics) Gauge(key string) gometrics.Gauge {
 		return gometrics.NilGauge{}
 	}
 
-	return s.registry.GetOrRegister(key, shm.NewShmGaugeFunc(s.fullName(key))).(gometrics.Gauge)
+	construct := func() gometrics.Gauge {
```
Comment (taoyuanyuan) [inline pkg/metrics/store.go:172]: same problem

File: pkg/metrics/store.go (hunk 2 above)
Comment (taoyuanyuan) [reply pkg/metrics/store.go:156]: NewLazyCounter直接传入Counter是不是更简单，后面的prefunc也就不需要了

File: pkg/metrics/store_lazy.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+// NewLazyCounter build lazyCounter
+func NewLazyCounter(ctor func() gometrics.Counter) (gometrics.Counter, error) {
+	if ctor == nil {
+		return nil, errors.New("build lazycounter ctor is empty")
+	}
+	return &lazyCounter{ctor: ctor}, nil
+}
+
+func (lc *lazyCounter) preFunc() {
```
Comment (taoyuanyuan) [inline pkg/metrics/store_lazy.go:24]: 同上

File: pkg/metrics/store.go (hunk 2 above)
Comment (taoyuanyuan) [reply pkg/metrics/store.go:156]: 哦 我知道了

**Review by taoyuanyuan — APPROVED** [review]
LGTM

**Review by nejisama — APPROVED** [review]

**Pull request merged**